src/nvim/buffer.c

```diff
@@ -1335,11 +1336,12 @@ void do_autochdir(void)
 static int top_file_num = 1;            /* highest file number */
 
 buf_T *
-buflist_new (
-    char_u *ffname,            /* full path of fname or relative */
-    char_u *sfname,            /* short fname or NULL */
-    linenr_T lnum,                  /* preferred cursor line */
-    int flags                      /* BLN_ defines */
+buflist_new(
+    char_u *ffname,            // full path of fname or relative
+    char_u *sfname,            // short fname or NULL
+    linenr_T lnum,             // preferred cursor line
+    int flags,                 // BLN_ defines
+    handle_T bufnr
```

Comment:
Hacking around, it looks like the code can be simplified a good amount by "not worrying" about the user accidentally editing this special buffer. I'd like to try to solve that more general problem rather than add the special cases for incsub buffer. There is already some internal notion in Vim for "special buffers", which we maybe can extend and formalize a bit instead of adding new parameters.

Tests are passing, I'll push a PR on top of this PR later today.